weasyl/oauth2.py

```diff
@@ -92,3 +118,72 @@ def get_userid_from_authorization(scopes=['wholesite']):
 __all__ = [
     'get_consumers_for_user', 'revoke_consumers_for_user',
 ]
+
+
+def get_allowed_scopes(user_id):
+    """
+    Get a list of oauth scopes this user is allowed to request
+    :param user_id: the userid of the application owner
+    :return: a list of scopes
+    """
+    allowed = _SCOPES
+    # only trusted individuals should be allowed to use the "wholesite" oauth grant
+    if user_id not in staff.ADMINS | staff.MODS | staff.DEVELOPERS:
+        allowed = [scope for scope in allowed if scope['name'] != 'wholesite']
+    return allowed
+
+
+def register_client(user_id, name, scopes, redirects, homepage):
+    """
+    Register an application as an OAuth2 consumer
+    :param user_id: the user registering this application
+    :param name: the name of the application
+    :param scopes: a list of the scopes registered for this application
+    :param redirects: allowed redirect URIs for this application
+    """
+    if not name.strip():
+        raise WeasylError("nameMissing")
+
+    session = d.connect()
+    new_consumer = orm.OAuthConsumer(
+        clientid=security.generate_key(32),
```

Comment:
@charmander probably has better opinions on how to do this properly than I, but since clientid isn't secret, could we make it something more meaningful? maybe a combination of the userid and the timestamp.